Add exercises for Git & Github Workshop Session 2 (#49)
WomenWhoCode/WWCode-London: resources/devops/Git & Github/Session2 - Exercises.md

Session 2  - Git and Github Exercises
---

Questions have been designed to be completed in the given sequence and is a continuation of exercises in [Session 1](Session1%20-%20Exercises.md)

> 1. Change directory to `git-workshop` and investigate the existing branches, commit history and status.

Reviewer: ```suggestion
1. Change directory to `git-workshop` and investigate the existing branches, the history of commits and status.
```

Author: `commit history` is a pretty standard terminology referenced across all of github documentation : https://docs.github.com/en/pull-requests/committing-changes-to-your-project/viewing-and-comparing-commits/differences-between-commit-views

Hence I will keep it like this.

    <details>
    <summary>Solution</summary>

    ```bash
    cd git-workshop
    git branch
    git log
    git status
    ```
    </details>

> 

Reviewer: Do we want to provide a solution for this step only?
I see 2 options:
- don't provide a solution at all
- provide solution to every step

Wdy?

Author: The intention is for volunteers to contribute to these solutions. I only added the first one as a sample format that can be used.

2. Create a new branch `session2` from the `session1` branch. Switch to the `session2` branch. Investigate where the HEAD is pointing to.
3. Add a new file `session2-file1` in the `session2` branch. Commit this new file to your local repository.
4. Edit the file `session2-file1` and add "SESSION2-RESTORE-STAGING" to the file. Add these changes to the staging area (do not commit the changes at this stage). Check the status of your local repository.
5. Unstage the changes in the file `session2-file1`. Check the status of your local repository.

> 6. Now Undo the changes in the file `session2-file1` and restore the file to HEAD. Check the status of your local repository. (Make sure that there are no changes to be committed)

Reviewer: ```suggestion
6. Undo the changes in the file `session2-file1` and restore the file to HEAD. Check the status of your local repository. Make sure that there are no changes to be committed.
```

7. Create 8 new commits in the branch `session2`. These can be any random commits that you can track easily. For simplicity you can use the format that was used during the session 2 event - Enter numerical values (ONE, TWO, THREE, etc) in the file `session2-file1` and commit after adding each value. Use commit messages that are easier to track ( Commit 1, Commit 2, Commit 3, etc). Check the commit history.

> 8. Undo the last 2 commits from the local repository such that the changes made are preserved in the working directory. Check git history and status. Undo the changes that exist in the working directory. (You should see 6 new commits in the log at this stage)

Reviewer: ```suggestion
8. Undo the last 2 commits from the local repository such that the changes made are preserved in the working directory. Check git history and status. Undo the changes that exist in the working directory. You should see 6 new commits in the log at this stage.
```

> 9. Now undo the latest 2 commits without preserving any changes made in these commits to the local directory. (You should only see 4 new commits in the log at this stage)

Reviewer: ```suggestion
9. Undo the latest 2 commits without preserving any changes made in these commits to the local directory. You should only see 4 new commits in the log at this stage.
```

> 10. Undo the latest 2 commits by preserving the git history that these commits were made and reverted.

Reviewer: Bonus exercise: `amend` commit and collapse all commits in 1?

Author: The exercises are based on what was covered during the session. We didn't cover it during this session.
